Added cancerBiomarkers data source

ireneisdoomed opened the pull request:
This PR processes the Cancer Biomarkers database available at `gs://otar000-evidence_input/CancerBiomarkers/data_files`.
The proposed schema has been tracked and can be observed in this spreadsheet: https://docs.google.com/spreadsheets/d/1Mowq7KsGTMtEg3wZpJBNK_UbawHKJeM9d0syT9F9AMc/edit#gid=613866016

#### datasourceId
- [X] `cancer_genome_interpreter` in favour of `cancer_biomarkers`. The latter rather describes the nature of the data.

#### datatypeId
- [X] `affected_pathway`. The data describes how the gene-drug interactions are altered when one or multiple variants are present.

#### diseaseFromSource
- [X] Split tumor types by ';' and explode

#### diseaseFromSourceMappedId
- [X] Extract EFOs by joining the tumor type with the disease mapping dataset
- [ ] This will eventually come via OnToma. The manual curation that cannot be automatised will be part of the manual disease mapping file which will be another input of OnToma.

#### drugFromSource
- This can be found in `Drug` (if the specific drug is reported) or `DrugFullName` (mainly the drug full name)
- [X] Coalesce both fields
- [X] Split `Drug` by ';' and explode

#### drugId
- [X] When the specific `Drug` is given and the mapping is straightforward, this field is joined with the latest drug index

#### drugResponse
- Possible values: 'Resistant', 'Not Responsive', 'Increased Toxicity', 'Increased Toxicity (Ototoxicity)', 'Increased Toxicity (Myelosupression)', 'Responsive', 'Increased Toxicity (Haemolytic Anemia)', 'Increased Toxicity (Hyperbilirubinemia)'
- [X] Mapped to EFO. 
- [ ] Open [issue to EFO](https://github.com/EBISPOT/efo/issues/1209) to: a) create terms in EFO for 'Resistant', 'Not Responsive' and 'Increased Toxicity'; b) have these terms grouped under `biological process`to describe responses to drug
- [X] This field is at the root level, not inside `biomarkers`. The described association is dependent on the presence of the biomarker, but the other relevant entities are also in the root so the relationship between them can be interpreted without the information encapsulated in biomarkers.

#### confidence
- Possible values:
```
+-------------------------------+-----+
|EvidenceLevel                  |count|
+-------------------------------+-----+
|Pre-clinical                   |520  |
|Early trials                   |298  |
|Case report                    |267  |
|FDA guidelines                 |128  |
|European LeukemiaNet guidelines|106  |
|NCCN guidelines                |64   |
|Late trials                    |47   |
|CPIC guidelines                |4    |
|Clinical trials                |2    |
|NCCN/CAP guidelines            |2    |
|Early Trials,Case Report       |2    |
|Late trials,Pre-clinical       |2    |
+-------------------------------+-----+
```
- [X] These will just be displayed but will not play a role for the scoring, as we know this dataset is the result of manual curation

#### literature
- [X] PubMed IDs are the source for 75% of the records.
- [X] Split by ';' and collected into an array.

#### urls
To source the other 25% without a PMID.
- [X] Extraction of the sources dataset where the nice names and urls are gathered.
- [X] Build a different structure whenever we have a CT: `niceName` will be the Clinical Trials and `url` is built using the NCT code.

#### targetFromSourceId
Every time that a biomarker consists of multiple variants that are not independent of each other, the biomarker is reported separating them with a '+'. 
When this situation happens, genes will be described under `Gene` separated by ';'. 
As we can only build evidence with a single target, these are separated into different evidence strings but the biomarker will reference both of them.
These cases account for 27 distinct biomarkers.
- [X] Split the Gene column by ';' and explode.
- [X] Correct some gene names to the official symbol

#### biomarkers
Array of structs that will capture dependent and independent variants, as well as secondary fields to describe the mutation.
The proposed fields are encapsulated in a struct so that a conceptual difference can be made when analysing data: `variantId` refers to a disease causing variant, whereas `biomarkers.variantId` adds the nuance of the biomarker having to be present for the association to occur.

#### biomarkers.name
- [X] Name of the whole string that describes the biomarker - including all variants

#### biomarkers.individualMutation
- [X] When multiple mutations are reported, this field indicates to which variant the whole record is pointing to.
- This field is populated only when the functional consequence of the variant is MUT.

#### biomarkers.variantFunctionalConsequenceId
- Possible values: 'MUT', 'CNA', 'FUS', 'EXPR', 'BIA'.
- [X] Map the types of alteration to an SO code
```
- MUT: somatic mutations --> SO:0001777
- CNA: copy number alterations --> SO_0001563
- FUS: fusion genes --> SO_0001882
- EXPR: mRNA expression --> SO_0001540
- BIA: biallelic inactivation --> not yet a proper term in SO. I've 
asked for its inclusion.
```
- [ ] Open [ticket to SO](https://github.com/The-Sequence-Ontology/SO-Ontologies/issues/548) to find or create a term for `BIA`.
- [X] This field will be under `biomarkers`, as there are cases where multiple variants are reported, each with a different consequence.

#### biomarkers.variantId
- [ ] Convert genomic coordinates to CHROM_POS_REF_ALT notation

#### biomarkers.variantRsId
- [ ] Investigate the use of the upcoming tool PepVEP to map the variantId to rsIDs

 #### biomarkers.variantAminoacidDescriptions
- [ ] Investigate the use of the upcoming tool PepVEP to extract the changes at the protein level
- NOTE: We would want to run this for all variants, not only cancer biomarkers

ireneisdoomed commented:
v2 of the parser brings significant changes, mainly:
- [X] Correct previous bugfixes
- [X] Adapt to the v2 of the schema - now open for PR ([#130](https://github.com/opentargets/json_schema/pull/130))
- [X] Rearrange code and adapt to PEP styles

**1321 evidence strings** have been generated and validated.
The next 3 exemplify how we will represent these 3 records shown in the data source:
![image](https://user-images.githubusercontent.com/45119610/130032526-49e76113-1e52-450b-b34d-c948a33cd49c.png)

 
```
-RECORD 0-----------------------------------------------------------
 biomarkers                | {null, ALK fusion, [SO_0001882], null}
 confidence                | Case report
 datasourceId              | cancer_genome_interpreter
 datatypeId                | affected_pathway
 diseaseFromSource         | Colorectal adenocarcinoma
 diseaseFromSourceMappedId | EFO_0000365
 drugFromSource            | Alk Inhibitors
 drugId                    | null
 drugResponse              | GO_0042493
 literature                | [26633560]
 targetFromSourceId        | ALK
 urls                      | null
 biomarker                 | ALK fusion
-RECORD 1-----------------------------------------------------------
 biomarkers                | {null, ALK fusion, [SO_0001882], null}
 confidence                | Case report
 datasourceId              | cancer_genome_interpreter
 datatypeId                | affected_pathway
 diseaseFromSource         | Colorectal adenocarcinoma
 diseaseFromSourceMappedId | EFO_0000365
 drugFromSource            | Ceritinib
 drugId                    | CHEMBL2403108
 drugResponse              | GO_0042493
 literature                | [26933125, 27742657, 26633560]
 targetFromSourceId        | ALK
 urls                      | null
 biomarker                 | ALK fusion
-RECORD 2-----------------------------------------------------------
 biomarkers                | {null, ALK fusion, [SO_0001882], null}
 confidence                | Case report
 datasourceId              | cancer_genome_interpreter
 datatypeId                | affected_pathway
 diseaseFromSource         | Colorectal adenocarcinoma
 diseaseFromSourceMappedId | EFO_0000365
 drugFromSource            | Entrictinib
 drugId                    | null
 drugResponse              | GO_0042493
 literature                | [26933125, 26633560]
 targetFromSourceId        | ALK
 urls                      | null
 biomarker                 | ALK fusion
```

Another example of how the variant information is modelled:
```
{
   "datasourceId":"cancer_genome_interpreter",
   "datatypeId":"affected_pathway",
   "drugFromSource":"Braf Inhibitors",
   "targetFromSourceId":"AKT1",
   "diseaseFromSource":"Cutaneous melanoma",
   "diseaseFromSourceMappedId":"EFO_0000389",
   "confidence":"Case report",
   "literature":[
      "24265152"
   ],
   "biomarkers":[
      {
         "name":"AKT1 (Q79K,E17K)",
         "individualMutation":"AKT1:Q79K",
         "variantId":"14_105243048_G_T",
         "variantFunctionalConsequenceIds":[
            "SO_0001777"
         ]
      },
      {
         "name":"AKT1 (Q79K,E17K)",
         "individualMutation":"AKT1:E17K",
         "variantId":"14_105246551_C_T",
         "variantFunctionalConsequenceIds":[
            "SO_0001777"
         ]
      }
   ]
}
```

DSuveges commented on modules/cancerBiomarkers.py:
```diff
@@ -0,0 +1,352 @@
+            'source', 'url')
+        disease_df = self.spark.read.json(disease_table).select(
+            regexp_replace(col('name'), '_', '').alias('tumor_type'),
+            regexp_extract(col('url'), '[^/]+$', 0).alias('diseaseFromSourceMappedId'))
```
See my comment on line 311
```suggestion
            regexp_extract(col('url'), r'[^/]+$', 0).alias('diseaseFromSourceMappedId'))
```

DSuveges commented on modules/cancerBiomarkers.py:
```diff
@@ -0,0 +1,352 @@
+        disease_df: DataFrame,
+        drugs_df: DataFrame
+    ) -> DataFrame:
+        """The diverse steps to prepare and enrich the input table are computed"""
```
```suggestion
        """The diverse steps to prepare and enrich the input table"""
```

DSuveges commented on modules/cancerBiomarkers.py:
```diff
@@ -0,0 +1,352 @@
+            .withColumn('gene', explode(col('gene')))
+            .withColumn('drug', explode(col('drug')))
+            # Override specific genes
+            .withColumn(
+                'gene',
+                when(col('gene') == 'C15orf55', 'NUTM1')
+                .when(col('gene') == 'MLL', 'KMT2A')
+                .when(col('gene') == 'MLL2', 'KMT2D')
+                .otherwise(col('gene'))
+            )
```
I think this is not necessarily the best idea to implement the mapping like this... I can imagine a growing set of genes with old or not canonical names and this pattern is not suitable. But this is very minor comment. And it's also quite deep down... a top level dictionary would be better like `ALTERATIONTYPE2FUNCTIONCSQ`. as you did in line 187 and 188.

ireneisdoomed replied:
I think it is not the same case, as, with the dictionary, the idea is that there is a mapping for every single value of the column, otherwise the column will be null. Whereas here we only want to filter and change this handful of gene names.

DSuveges commented on modules/cancerBiomarkers.py:
```diff
@@ -0,0 +1,352 @@
+                    col('alteration').contains(':.'),
+                    translate(col('alteration'), ':.', '')
+                )
+                .when(
+                    col('alteration') == 'NRAS:.12.,.13.,.59.,.61.,.117.,.146.',
+                    col('Biomarker')  # 'NRAS (12,13,59,61,117,146)'
+                )
```
I think if `col('alteration').contains(':.')` in row 143, this condition will never be satisfied.

DSuveges commented on modules/cancerBiomarkers.py:
```diff
@@ -0,0 +1,352 @@
+                'literature',
+                when(col('source').startswith('PMID'), regexp_extract(col('source'), r'(PMID:)(\d+)', 2))
+            )
+            .withColumn(
+                'urls',
+                when(
+                    col('source').startswith('NCT'),
+                    struct(
+                        lit('Clinical Trials').alias('niceName'),
+                        concat(lit('https://clinicaltrials.gov/ct2/show/'), col('source')).alias('url')
+                    )
+                )
+                .when(
+                    (~col('source').startswith('PMID')) | (~col('source').startswith('NCIT')),
+                    struct(col('niceName'), col('url'))
+                )
+            )
```
This is really neatly done. I like it.

ireneisdoomed commented on modules/cancerBiomarkers.py:
```diff
@@ -0,0 +1,352 @@
+        Zips in a tuple the combination of the alteration w/ its correspondent type 
+        so that when multiple alterations are reported, these can be disambiguated.
+        By expanding the array of alteration types it accounts for the cases when
+        several alterations are reported but only one type is given
```
Thanks. Agree it is not good to mess concepts. I've changed the names:
- `main` is now `compute`
- `compute_biomarkers` is now `process_biomarkers`

ireneisdoomed commented on modules/cancerBiomarkers.py:
```diff
@@ -0,0 +1,352 @@
+            ArrayType(ArrayType(StringType()))
+        )
+
+    def main(
```
Thanks. Agree it is not good to mess concepts. I've changed the names:
- `main` is now `compute`
- `compute_biomarkers` is now `process_biomarkers`

ireneisdoomed commented:
Changes from the previous version:

- Completed all TODOs:
- [x] Some evidence strings currently do not pass validation due to some inaccuracies in the parsing of the variantId.
- [x] This is one string in drugFromSource: `[tamoxifen,letrozole,anastrozole,exemestane,fulvestrant,lhrh Analogues Or Antagonist]`. I should clean it and explode it.
- [x] There are 6 evidence strings in which `confidence` contains two levels of confidence. Split these by comma and explode the evidence strings. 
- Changed the name of the data source to `cancer_biomarkers` to distinguish it from the other tools from Cancer Genome Interpreter.
- `drugResponse` mapping to EFO is switched off for the moment until we hear back from EFO. It now consists of the raw string.
- Addressed @DSuveges comments.
